docs: Add jsdoc comments for driver functions

hjetpoluru opened the pull request:
## **Description**

As part of our documentation efforts, adding JSDoc comments to the driver function will enhance understanding of what the functions perform. Including examples for some functions is essential for better comprehension and usage, leading to improved coverage. This may also help in reducing flakiness and improving code consistency. Anyone onboarding and wishing to use automation with a customized Selenium driver should be able to manage effectively.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23699?quickstart=1)

## **Related issues**

Fixes: #23698 

## **Manual testing steps**

[Click here to access README ](https://github.com/MetaMask/metamask-extension/blob/doc/add-jsdoc-comments/test/e2e/webdriver/README.md)

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

hjetpoluru committed:
Added jsdoc comments for the driver methods

hjetpoluru committed:
Fixed lint

brad-decker commented on test/e2e/webdriver/driver.js:
```diff
@@ -149,6 +149,29 @@ class Driver {
+   * To target an element based on its attribute using a CSS selector, use square brackets ([]) to specify the attribute name and its value.
+   * await driver.findElement('[data-testid="eth-overview-buy"]'); // Syntax for CSS selector by data-testid(HTML attribute and value)
+   *
+   * To locate an element by XPath, use a path expression to navigate through elements and attributes in the HTML document.
+   * await driver.findClickableElement({ text: 'Confirm', tag: 'button' }); // Syntax for locating the button element that contains text ‘Confirm’
```
Is @example typically under @returns/@throws ?

hjetpoluru commented on test/e2e/webdriver/driver.js:
```diff
@@ -149,6 +149,29 @@ class Driver {
   }
 
+  /**
+   * Builds a locator object for the given element. This function exemplifies a sophisticated
+   * approach to element location, catering to various scenarios and preferences in locator
+   * specification. By encapsulating multiple locator strategies into a single, unified
+   * interface, it significantly streamlines the test writing process.
```
Ok thanks for the feedback looks like the language isn't clear, surely I'll simplify it. Understanding this method is crucial because it's the locator strategy we use to identify elements. When applied correctly, it could significantly reduce flakiness. I will do my best to represent it.

Stam098c replied:
[pro_gnupg_de.pdf](https://github.com/MetaMask/metamask-extension/files/14972839/pro_gnupg_de.pdf)

hjetpoluru committed:
Addressed review comments

hjetpoluru committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

hjetpoluru committed:
Added README section to explain driver functionality - WIP

hjetpoluru committed:
Added the elements about the driver - WIP

hjetpoluru committed:
Element and interactions - WIP

hjetpoluru committed:
Fixed the image referrence using relative path

hjetpoluru committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

hjetpoluru committed:
Waits and actions - completed

hjetpoluru committed:
Completed

hjetpoluru committed:
Fixed some format issues

hjetpoluru committed:
Fixed lint

hjetpoluru committed:
Added line breaks

hjetpoluru committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

hjetpoluru committed:
Improved the representation for locators

hjetpoluru commented:
@HowardBraham , Thank you for the feedback. I've simplified and updated the text, including the parameter descriptions and return values as you suggested. I've also provided detailed explanations in the README and organized it accordingly. Please let me know if there's anything else that needs to be done. 
Note:- I have included examples for the essential functions in the Driver, and all are covered in the README.

hjetpoluru committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

HowardBraham commented on test/e2e/webdriver/README.md:
````diff
@@ -0,0 +1,1228 @@
+>>
+</details>
+</details>
+<details><summary><b>Locate element by XPath</b></summary>
+
+>
+>To locate an element by XPath,to navigate through elements and attributes in the HTML document.
+>
+>![Screenshot displays ‘Confirm’ button in the send transaction page](image/xpath.png)
+>
+>Syntax for locating the button element that contains text ‘Confirm’
+>
+>```tsx
+>await driver.findClickableElement({ text: 'Confirm', tag: 'button' });
+>```
+>
+>Another syntax for locating the div menu element that contains text ‘Settings’
+>
+>```tsx
+>await driver.clickElement({ text: 'Settings', tag: 'div' });
+>```
+>>
+</details>
````
In this example, called "Locate element by XPath," you're showing how to locate by Text and Tag (I know under the hood it becomes an XPath)

hjetpoluru committed:
embedded video related to xpath

HowardBraham committed:
using GitHub video

hjetpoluru committed:
Addressed Howard comment on XPath

hjetpoluru committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

hjetpoluru committed:
Included the new functions

hjetpoluru committed:
Corrected all source links as per feedback from Chloe

hjetpoluru committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

hjetpoluru committed:
Fixed lint

HowardBraham committed:
improving the XPath section

seaona commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+
+Assertion is a statement that checks if a specified condition is true. If the condition is true, the program continues to execute. If the condition is false, throws an error or fails.
+
+# When do we need Assertions?
```
```suggestion
### When do we need Assertions?
```

seaona commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+- Ensuring that a web element is visible, contains specific text, or is enabled/disabled.
+- Verify that a certain condition holds at a specific point in the program or test case.
+
+# key points to remember
```
```suggestion
### Key points to remember
```

seaona commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+
+Selenium provides several mechanisms for managing waits, crucial for handling the asynchronous nature of web applications where elements may load at different times. These waits help avoid errors in test scripts that occur when trying to interact with elements that are not yet available on the web page
+
+## When do we need to wait?
```
```suggestion
### When do we need to wait?
```

seaona commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+
+### \***\* NOTE - Delay \*\***
+
+**`delay`** function is hard-coded wait may be longer than needed, resulting in slower test execution. Please avoid using this function.
```
```suggestion
** The `delay`** function is a hard-coded wait that may be longer than needed, resulting in slower test execution. Please avoid using this function.
```

seaona commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+
+- **Going to a URL**: This is the starting point of navigation, where you direct the browser to load a specific web page by providing its URL.
+- **Refreshing the Page**: Refreshes the current page to update content or re-execute certain actions.
+- **Handling Redirects**: Web pages might redirect to different URLs automatically. Automation scripts needs to handle these redirects gracefully, ensuring they can continue to interact with the correct page elements after the redirect.
```
```suggestion
- **Handling Redirects**: Web pages might redirect to different URLs automatically. Automation scripts need to handle these redirects gracefully, ensuring they can continue to interact with the correct page elements after the redirect.
```

legobeat commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+
+## buildLocator
+
+**`buildLocator`** function enhances element matching capabilities by introducing support for inline locators, offering an alternative to the traditional use of Selenium's By abstraction.
```
```suggestion
The **`buildLocator`** function enhances element-matching capabilities by introducing support for inline locators, offering an alternative to the traditional use of Selenium's `By` abstraction.
```

legobeat commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+
+[source](https://github.com/MetaMask/metamask-extension/blob/1f2bfb388695034db8859877ed21b4b045514f9f/test/e2e/webdriver/driver.js#L190)
+
+**Arguments**<br>
```
```suggestion
**Arguments**
```
There's a fair amount of `<br>`s here - are they necessary in this markdown section?

legobeat commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+<details><summary><b>findElement</b></summary>
+<br>
+
+> **`findElement`** function is called on the driver instance, it returns a reference to the first element in the DOM that matches with the provided locator. This value can be stored and used for future element actions.
```
```suggestion
> **`findElement`** returns a reference to the first element in the DOM thatthe provided locator matches. 
```

TBH I am not sure what `This value can be stored and used for future element actions.` is supposed to mean - if it's worth keeping, maybe elaborate with an example?

legobeat commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1396 @@
+
+## Interactions
+
+When we talk about interacting with web elements, we mean simulating the actions a user might take on a webpage. This includes:
```
```suggestion
In this context, interacting with web elements implies simulating actions a user might take on a webpage. This includes:
```

hjetpoluru committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

hjetpoluru committed:
Addressed review comments made by Mariona

hjetpoluru committed:
Addressed the comments made by legoBeat

hjetpoluru committed:
Updated the formatting by removing the br

hjetpoluru committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

hjetpoluru committed:
Fixed lint

hjetpoluru committed:
Fixed lint

hjetpoluru committed:
Merge branch 'develop' into doc/add-jsdoc-comments

legobeat commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1309 @@
+
+In web automation testing, locators are crucial commands that guide the framework to identify and select [HTML elements](https://www.w3schools.com/html/default.asp) on a webpage for interaction. They play a vital role in executing various actions such as clicking buttons, fill text, or retrieving data from web pages. Gaining a solid understanding of locators is a key step in initiating web testin ...
+
+## buildLocator
```
```suggestion
### buildLocator
```
Heading level

legobeat commented on test/e2e/webdriver/README.md:
```diff
@@ -0,0 +1,1309 @@
+>
+> [source](https://github.com/MetaMask/metamask-extension/blob/671c9975424a83904a4752dfb8a7cf728ae67355/test/e2e/webdriver/driver.js#L236)
+>
+> **Arguments**
```
Style

hjetpoluru committed:
Update test/e2e/webdriver/README.md

Co-authored-by: legobeat <109787230+legobeat@users.noreply.github.com>

hjetpoluru committed:
Style updated

hjetpoluru committed:
Merge branch 'develop' into doc/add-jsdoc-comments

hjetpoluru committed:
Merge branch 'develop' into doc/add-jsdoc-comments

hjetpoluru committed:
Merge branch 'develop' into doc/add-jsdoc-comments

hjetpoluru committed:
Updated the document to fix the type error raised for switchToWindowWithTitle

hjetpoluru committed:
Merge branch 'doc/add-jsdoc-comments' of github.com:MetaMask/metamask-extension into doc/add-jsdoc-comments

hjetpoluru committed:
Merge branch 'develop' into doc/add-jsdoc-comments

hjetpoluru committed:
Merge branch 'develop' into doc/add-jsdoc-comments

legobeat committed:
Merge branch 'develop' into doc/add-jsdoc-comments

HowardBraham committed:
Merge branch 'develop' into doc/add-jsdoc-comments
